lessons/specifics/guardian.md

```diff
@@ -0,0 +1,264 @@
+
+{% include toc.html %}
+
+## JWTs
+
+JWT provides a rich token for authentication. Where many authentication systems provide access to only a subject identifier for the resource, JWTs provide this along with other information like:
+
+* Who issued the token
+* Who is the token for
+* Which system should use the token
+* What time was it issued
+* What time should you consider it valid
+* What time does the token expire
+
+In addition to these fields Guardian provides some other fields to facilitate additional functionality.
+
+* What type is the token
+* What permissions does the bearer have
+
+These are just the basic fields in a JWT. You're free to add any additional information that your application requires. Just remember to keep it short. The JWT has to fit in the HTTP header.
+
+This richness means that you can pass JWTs around in your system as a fully contained unit of credentials. 
+
+### Where to use them
+
+JWT tokens can be used to authenticate any part of your application. 
+
+* Single page applications
+* Controllers (via browser session)
+* Controllers (via authoirzation headers - API)
+* Phoenix Channels
+* Service to Service requests
+* Inter-process
+* 3rd Party access (OAuth)
+* Remember me functionality
+* Other interfaces - raw TCP, UDP etc
+* Command line (if you _really_ want)
+
+JWT tokens can be used everywhere in your application where you need to provide verifiable authentication.
+
+### Do I have to use a database
+
+You do not need to track JWT via a database. You can simply rely on the issued and expiry timestamps for controlling access. Often you'll end up using a database to look up your user resource but the JWT itself does not require it.
+
+For example, if you were going to use JWT to authenticate communication on a UDP socket you likely wouldn't use a database. Encode all the information you need directly into the token when you issue it. Once you verify it (check that it's signed correctly) you're good to go.
+
+You _can_ however use a database to track JWT. If you do, you gain the ability to verify that the token is still valid - that is - it has not been revoked. Or you could use the records in the db to force a logout of all tokens for user 5. This is made simple in Guardian by using [GuardianDb](https://github.com/hassox/guardian_db). GuardianDb uses Guardians 'Hooks' to perform validation checks, save and delete from the db. We'll cover that later.
+
+### Can I use someone elses token?
+
+Often folks wonder if it's appropriate to use an OAuth token from Facebook or Google as their authentication token. It is not. The behaviours of those tokens, their validity is defined outside your application. There's a more security focused reason though. Anyone who obtains the token not only has access to your site, they also have access to Facebook/Google/Some Provider. 
+
+Always use your own tokens, other applications tokens are useful to identify someone and make requests against their systems.
```

Comment:
Ok, that makes sense. It would be good if there was an explanation of the whole guardian / ueberauth project and sub project relationships at the start then because that could be confusing.